Code refactoring, including changing filters to jinja tests

raphendyr opened the pull request:
Iterated over the code...

Most important are the first two commits. First will allow jinja to use cached version of field_macros as the context is not exposed (there is no need to do that). Second makes the templates way more clear by using jinta test syntax instead of django like filters. In addition it doesn't pollute filter namespace with strange filters.

Last two commits rewrite render and bootstrap_horizontal.

In addition, I added some comments :smile:

raphendyr committed:
Allow jinja to cache field_macros by not exposing context

raphendyr committed:
Convert field test filters to jinja tests

In addition, removed extra space when not multiple checkbox field

raphendyr committed:
Refactor render function

 * Don't require markup_classes
 * Accept partially populated markup_classes (provide defaults)
 * Replace strange test for bound field with inheritance check
 * Break if .. else { if .. else } to if .. elif .. else

raphendyr committed:
Reimplement bootstrap_horizontal to look less of a hack

 * can take label_cols as list or a string
 * can take max_columns (defaults to BOOTSTRAP_COLUMN_COUNT)
 * Fix space and order in test fixtures

samuelcolvin commented:
tests seem to be failing.

samuelcolvin commented on bootstrapform_jinja/templatetags/bootstrap.py:
```diff
@@ -1,110 +1,148 @@
-from django import forms
+from django.forms import CheckboxInput, CheckboxSelectMultiple, \
```
I don't think we document it here, but we always use 120 max line length and parenthesis when imports span multiple lines/

raphendyr replied:
I prefer parenthesis my self too, but used that format for some reason. I'll check the 120 limit too.

samuelcolvin commented on bootstrapform_jinja/templatetags/bootstrap.py:
```diff
@@ -1,110 +1,148 @@
 from django.template.loader import get_template
-from django_jinja import library
 from django.utils.safestring import mark_safe
+from django.utils.six import string_types
```
I have a policy that all code written or edited since 1st Jan 2017 is python 3.5+ only. Is that ok with you?

raphendyr replied:
It's. Did that only as I wasn't sure and it wasn't too complicated.

samuelcolvin commented on bootstrapform_jinja/templatetags/bootstrap.py:
```diff
@@ -1,110 +1,148 @@
+    spaces when the set is converted to string.
+    """
+    def __str__(self):
+        return ' '.join((str(x) for x in self))
```
no need for extra parenthesis `return ' '.join(str(x) for x in self)`

samuelcolvin replied:
or even just `' '.join(map(str, self))`

raphendyr replied:
If I remember right python 2.7 required extra parenthesis, but as that doesn't need to be supported those could be dropped for sure. Note to self: remember that map function exists...

samuelcolvin commented on bootstrapform_jinja/templatetags/bootstrap.py:
```diff
@@ -1,110 +1,148 @@
+from bootstrapform_jinja.config import BOOTSTRAP_COLUMN_COUNT
+
+
+class ClassSet(set):
```
`SpaceSet` would be a more sensible name I think.

raphendyr replied:
Probably. Less s letters at least :)

samuelcolvin commented on bootstrapform_jinja/templatetags/bootstrap.py:
```diff
@@ -1,110 +1,148 @@
-    for cl in label_cols.split(' '):
-        split_class = cl.split('-')
-
+    for cl in label_cols:
```
don't think I wrote this originally, it came from django bootstrap form. I'll believe you that it works if we have some proper tests and coverage.

raphendyr replied:
Yeah. It clearly was just copy from the original source. Though, I didn't have interest of rewriting that. In perfect world both codes would use same implementation. I'm most afraid that I made some mistake when doing this part.

samuelcolvin reviewed (changes requested):
In general I think this looks good, but there needs to be a bit of work done before it's ready to merge.

We've been using django-jinja-bootstrap-form in production a lot for years and it's "just worked" so we need to make absolutely sure it's functionality hasn't altered before releasing an update.

If you get tests passing and fix the majority of the commends, I'll merge then improve tests, coverage, linting etc.

raphendyr commented:
I fixed the test framework and only difference in horizontal is different order of classes (due to use of Set). Apparently `assertHTMLEqual` compares attributes as strings and doesn't handle class attribute specially. I was planning of using `' '.join(sorted(map(str, self)))` to sort this out.

Any case, I'll finalize this stuff tomorrow.

samuelcolvin commented:
> I was planning of using ' '.join(sorted(map(str, self))) to sort this out.

LGTM

raphendyr committed:
Update tests for django 1.9 and up

raphendyr commented:
Removed usage of the set and just made sure the input doesn't contain duplicates (and is strings, which would have broken the for loop any case). The code should be bit more clear now.

I changed the tests a bit to work in django 1.9 and up (version test didn't work correctly). I also replace render_to_response with render (as previous is deprecated). Only change in the template output is spaces in class attributes, so nothing in functionality should have changed.

raphendyr committed:
Fix tests for newer django versions and test those in travis

raphendyr committed:
Ensure that template tags are loaded to jinja library

raphendyr commented:
Tests where failing in travis as with django 1.8 and 1.7 django-jinja wasn't importing templatetags for some reason. Now AppConfig file will do it in the ready method (supported in django 1.7). I also fixed tests up to django 1.11 and added those to travis.

samuelcolvin commented:
Look good initially, I'll have a proper look tomorrow.

By the way before the next release I'll drop support for python <3.5 and django <1.8 and anything before the latest django-jinja. Hope that's ok with you.

raphendyr commented:
Should be fine.

samuelcolvin commented:
thanks a lot.
